### PR TITLE
ENT-8430: Fixed default for ignore_missing_bundles and ignore_missing_inputs 

### PR DESCRIPTION
**Pull request by olehermanse**
I fixed this issue, but refrained from starting to fix more than necessary.

Merge together:
https://github.com/cfengine/core/pull/4929
https://github.com/cfengine/masterfiles/pull/2268

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom please build this in jenkins.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7649/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7649/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7649/

**Comment by olehermanse** [comment]
@cf-bottom please build this in jenkins, now that I refactored the functions a bit.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7651/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7651/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7651/

**Comment by larsewi** [comment]
Why introduce `StringIsBoolean` in one commit, and change it in another commit? Why not introduce the final version first?

**Comment by olehermanse** [comment]
@larsewi on a serious note. I did the work (code changes) in that order, and explained what I was doing in each commit. Now, for example if you disagree with the simplifications I did in the last commit, we can drop that one. If on the other hand you think it would be better that I squashed it with the first one, I can do that as well. (In some situations like this, squashing / reordering commits would be difficult as it can introduce conflicts). Time travel (rebase) is tricky business.

**Comment by larsewi** [comment]
Yes, we have the technology to time travel. Rebase is a time machine ⌛

**Comment by larsewi** [comment]
No, I'm fine with it. I ask more out of curiosity

**Review by larsewi — APPROVED** [review]
ACK

File: libpromises/expand.c [hunk 1]
```diff
@@ -962,16 +962,22 @@ static void ResolveControlBody(EvalContext *ctx, GenericAgentConfig *config,
         {
             Log(LOG_LEVEL_VERBOSE, "SET ignore_missing_inputs %s",
                 RvalScalarValue(evaluated_rval));
-            config->ignore_missing_inputs = BooleanFromString(
-                RvalScalarValue(evaluated_rval));
+            if (StringIsBoolean(RvalScalarValue(evaluated_rval)))
+            {
+                config->ignore_missing_inputs = BooleanFromString(
```
Comment (vpodzime) [inline libpromises/expand.c:967]: So this is initialized to `false`?

**Review by vpodzime — COMMENTED** [review]
I'd personally add a static array of strings to compare to instead of (or together with) the `CF_BOOL` comma-separated list of words as one string. But YMMV.

File: libpromises/expand.c (hunk 1 above)
Comment (olehermanse) [reply libpromises/expand.c:967]: Yes, it is initialized to false and documented that false is the default. We may have to add some tricks, either to the tests, or to the C code, to make it default to false at a later point, for example after pre-eval or after some evaluation passes. Need to test more.

**Comment by olehermanse** [comment]
> I'd personally add a static array of strings to compare to instead of (or together with) the CF_BOOL comma-separated list of words as one string. But YMMV.

@vpodzime I don't really disagree, but I left CF_BOOL like that since it's following a similar format to other macro strings (syntax options for various attributes).

**Review by craigcomstock — DISMISSED** [review]

**Comment by vpodzime** [comment]
> I left CF_BOOL like that since it's following a similar format to other macro strings (syntax options for various attributes)

Yeah, I saw that (later). But having that array-of-strings representation next to it wouldn't hurt. Maybe one day... :wink:

**Comment by olehermanse** [comment]
@cf-bottom please test this in Jenkins :)

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7674/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7674/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7674/

**Review by vpodzime — COMMENTED** [review]
Please create a follow-up ticket for the added `config` -- I guess we can now remove some things from the `EvalContext`. Also, the *Fixed inconsistencies with methods...* commit seems to be removing an early check that prevented partial evaluation of the policy in case of missing bundles. Do we really want to drop that?

**Comment by olehermanse** [comment]
@cf-bottom Please test this in jenkins now that I added more tests.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7677/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7677/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7677/

**Comment by olehermanse** [comment]
> Please create a follow-up ticket for the added config -- I guess we can now remove some things from the EvalContext. Also, the Fixed inconsistencies with methods... commit seems to be removing an early check that prevented partial evaluation of the policy in case of missing bundles. Do we really want to drop that?

In short, yes. I have tried to improve the commit message:

https://github.com/cfengine/core/pull/4929/commits/ee10942c94e606a6d1fdb9be6333c59965748094

I created 2 follow-up tickets:

https://tracker.mender.io/browse/CFE-3940
https://tracker.mender.io/browse/CFE-3941

I also added appropriate tests:

https://github.com/cfengine/core/pull/4929/commits/57f2ce6257421adfefab31b00aa5dd231100bb91

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7680/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7680/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7680/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7681/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7681/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7681/

File: tests/acceptance/01_vars/02_functions/mustach_doesnt_render_null.cf [hunk 2]
```diff
@@ -31,6 +31,7 @@ bundle agent check
 
       "check" usebundle => dcs_check_regcmp("one two three ", $(test.mustache_string), $(this.promise_filename), "no");
 
+  reports:
```
Comment (nickanderson) [inline tests/acceptance/01_vars/02_functions/mustach_doesnt_render_null.cf:34]: Yeah, I saw this one today as well. Bizarre.

**Review by nickanderson — APPROVED** [review]
Tests make sense to me.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7683/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7683/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7683/

**Review by nickanderson — DISMISSED** [review]
Policy tests look OK to me. Might be worth having a version where the bundle is defined after it's use.

File: libpromises/conversion.c [hunk 3]
```diff
@@ -351,80 +351,41 @@ DataType ConstraintSyntaxGetDataType(const ConstraintSyntax *body_syntax, const
 
 /****************************************************************************/
 
+// Warning: Defaults to true on unexpected (non-bool) input
 bool BooleanFromString(const char *s)
```
Comment (vpodzime) [inline libpromises/conversion.c:355]: These two functions could now be `inline`, but no biggie.

**Review by vpodzime — DISMISSED** [review]
Looks good to me. Thanks for the extra work on this! :clap:

File: tests/acceptance/21_methods/reachable/undefined_bundle_promiser.x.cf [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# These tests are left very simple on purpose,
+# they are not using the test framework.
+# They should be seen together, note that defined vs undefined variants are
+# almost identical, just replacing one detail to show what should work
+# and what should error.
+
+bundle agent main
+{
+  methods:
+    "foo"; # Should error because bar does not exist
```
Comment (larsewi) [inline tests/acceptance/21_methods/reachable/undefined_bundle_promiser.x.cf:10]: s/bar/foo/g

File: tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  reports:
+    "$(x)";
+    "$(this.promise_filename) Pass";
+}
+
+bundle agent main
+{
+  methods:
+    "foo"
+      usebundle => bar("baz");
```
Comment (larsewi) [inline tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf:18]: This should error because `baz` does not exist? Maybe add a comment?

Comment (larsewi) [inline tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf:18]: I don't understand this corner case. Maybe a comment would help ;)

File: libpromises/conversion.c (hunk 3 above)
Comment (larsewi) [reply libpromises/conversion.c:355]: Maybe we should add some rules on when to inline in [CONTRIBUTING.md](https://github.com/cfengine/core/blob/master/CONTRIBUTING.md)?

Comment (vpodzime) [reply libpromises/conversion.c:355]: It's part of the craft. :wink:

File: tests/acceptance/21_methods/reachable/undefined_bundle_promiser.x.cf (hunk 4 above)
Comment (olehermanse) [reply tests/acceptance/21_methods/reachable/undefined_bundle_promiser.x.cf]: Thanks, fixed.

File: tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf (hunk 5 above)
Comment (olehermanse) [reply tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf]: I could add a comment but it would say the same as it says in the top of this file. This test doesn't make sense on it's own, but when you look at it together with the other test with undefined in the name it makes a bit more sense. The one where it's defined should pass, and the one where it's undefined should fail. They are identical except changing that one bundle name.

**Comment by olehermanse** [comment]
@cf-bottom please test this in jenkins, now that I linked it to Vratas PR.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7705/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7705/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7705/

File: tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# These tests are left very simple on purpose,
+# they are not using the test framework.
+# They should be seen together, note that defined vs undefined variants are
+# almost identical, just replacing one detail to show what should work
+# and what should error.
+
+bundle agent bar(x)
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply tests/acceptance/21_methods/reachable/defined_bundle_usebundle_parameters.cf]: I have added a couple of tests which do this. (But I think it's kinda off topic here).

**Review by vpodzime — APPROVED** [review]
ACK

**Comment by olehermanse** [comment]
3.18.x cherry-pick:

https://github.com/cfengine/core/pull/4934
